### PR TITLE
CORE-8882: MTLS - Check client certificate

### PR DESCRIPTION
**Pull request by yift-r3**
This pull-request checks if the client certificate is allowed (i.e., added to the allowed list) before validating it.

### CODE REVIEW

File: components/gateway/src/main/kotlin/net/corda/p2p/gateway/messaging/http/DynamicX509ExtendedTrustManager.kt [hunk 1]
```diff
@@ -65,17 +56,52 @@ internal class DynamicX509ExtendedTrustManager(
...
     override fun getAcceptedIssuers(): Array<X509Certificate> {
         // We assume here that all the trust stores were issued by the same CA.
-         return getAllX509TrustManagers().map { it.acceptedIssuers.toList() }.flatten().toTypedArray()
+        return getAllX509TrustManagers().flatMap {
+            it.acceptedIssuers.toList()
+        }.toTypedArray()
+    }
+
+    @Suppress("ThrowsCount")
+    private fun validateClientCertificateChange(chain: Array<out X509Certificate>?) {
```
Comment (williamvigorr3) [inline components/gateway/src/main/kotlin/net/corda/p2p/gateway/messaging/http/DynamicX509ExtendedTrustManager.kt]: `validateClientCertificateChain`?

File: components/gateway/src/test/kotlin/net/corda/p2p/gateway/messaging/mtls/DynamicCertificateSubjectStoreTest.kt [hunk 2]
```diff
@@ -53,35 +55,79 @@ class DynamicCertificateSubjectStoreTest {
...
+        assertThat(dynamicCertificateSubjectStore.subjectAllowed(MemberX500Name.parse(SUBJECT_1))).isTrue
 
         processor.firstValue.onNext(Record(TOPIC, KEY_2, null), ClientCertificateSubjects(SUBJECT_1), emptyMap())
 
-        assertThat(dynamicCertificateSubjectStore.subjectAllowed(SUBJECT_1)).isFalse
+        assertThat(dynamicCertificateSubjectStore.subjectAllowed(MemberX500Name.parse(SUBJECT_1))).isFalse
     }
 
+    @Test
+    fun `onNext will add the normalizes subject`() {
```
Comment (williamvigorr3) [inline components/gateway/src/test/kotlin/net/corda/p2p/gateway/messaging/mtls/DynamicCertificateSubjectStoreTest.kt]: `normalizes` -> `normalized` (and on line 107).

File: components/gateway/src/main/kotlin/net/corda/p2p/gateway/messaging/mtls/DynamicCertificateSubjectStore.kt [hunk 3]
```diff
@@ -42,14 +43,16 @@ internal class DynamicCertificateSubjectStore(
...
+        val normalizedSubject = MemberX500Name.parse(subject)
+        certificateSubjects.compute(normalizedSubject) { _, sources ->
             sources?.apply { add(source) } ?: mutableSetOf(source)
         }
     }
 
-    private fun removeSource(subject: CertificateSubject, source: String) {
-        certificateSubjects.computeIfPresent(subject) { _, sources ->
+    private fun removeSource(subject: String, source: String) {
+        val normelizedSubject = MemberX500Name.parse(subject)
```
Comment (williamvigorr3) [inline components/gateway/src/main/kotlin/net/corda/p2p/gateway/messaging/mtls/DynamicCertificateSubjectStore.kt]: `normelizedSubject` -> `normalizedSubject`

**Review by williamvigorr3 — COMMENTED** [review]
Looks good. Left some minor comments.

**Review by williamvigorr3 — APPROVED** [review]
LGTM

**Review by dimosr — APPROVED** [review]

**Pull request merged**